server/market/market.go

```diff
@@ -574,10 +610,170 @@ func (m *Market) processOrder(ctx context.Context, rec *orderRecord, epoch *Epoc
+		return
+	}
+	err = json.Unmarshal(resp.Result, &piResp)
+	if err != nil {
+		sendPI(nil)
+		m.respondError(msg.ID, reqData.user, msgjson.RPCParseError,
+			fmt.Sprintf("error parsing preimage notification response payload result: %v", err))
+		return
+	}
+
+	// Validate preimage length.
+	if len(piResp.Preimage) != order.PreimageSize {
+		sendPI(nil)
+		m.respondError(msg.ID, reqData.user, msgjson.InvalidPreimage,
+			fmt.Sprintf("invalid preimage length (%d byes)", len(piResp.Preimage)))
+		return
+	}
+
+	// Check that the preimage is the hash of the order commitment.
+	var pi order.Preimage
+	copy(pi[:], piResp.Preimage)
+	piCommit := pi.Commit()
+	if reqData.commit != piCommit {
+		sendPI(nil)
+		m.respondError(msg.ID, reqData.user, msgjson.PreimageCommitmentMismatch,
+			fmt.Sprintf("preimage hash %x does not match order commitment %x",
+				piCommit, reqData.commit))
+		return
+	}
+
+	// The preimage is good.
+	log.Tracef("Good preimage received for order %v: %x", reqData.id, pi)
+	sendPI(&pi)
+}
+
+// collectPreimages solicits preimages from the owners of each of the orders in
+// the provided queue with a 'preimage' ntfn/request via AuthManager.Request,
+// and returns the preimages contained in the client responses. This function
+// can block for up to 5 seconds (piTimeout) to allow clients time to respond.
+// Clients that fail to respond, or respond with invalid data (see
+// handlePreimageResp), are counted as misses.
+func (m *Market) collectPreimages(orders []order.Order) (cSum []byte, ordersRevealed []*matcher.OrderRevealed, misses []order.Order) {
+	// Compute the commitment checksum for the order queue.
+	cSum = matcher.CSum(orders)
+
+	// Request preimages from the clients.
+	piTimeout := 5 * time.Second
+	preimages := make(map[order.Order]chan *order.Preimage, len(orders))
+	for _, ord := range orders {
+		// Make the 'preimage' request.
+		piReqParams := &msgjson.PreimageRequest{
+			OrderID:        idToBytes(ord.ID()),
+			CommitChecksum: cSum,
+		}
+		req, err := msgjson.NewRequest(comms.NextID(), msgjson.PreimageRoute, piReqParams)
+		if err != nil {
+			// This is likely an impossible condition, but it's not the client's
+			// fault.
+			log.Errorf("error creating preimage request: %v", err)
+			// TODO: respond to client with server error.
+			continue
+		}
+
+		// The clients preimage response comes back via a channel, where nil
+		// indicates client failure to respond, either due to disconnection or
+		// no action.
+		piChan := make(chan *order.Preimage)
+
+		reqData := &piData{
+			user:     ord.User(),
+			id:       ord.ID(), // maybe remove
+			commit:   ord.Commitment(),
+			preimage: piChan,
+		}
+
+		// Failure to respond in time is a miss, signalled by a nil pointer.
+		miss := func() {
+			piChan <- nil
+		}
+
+		// Send the preimage request to the order's owner.
+		err = m.auth.Request(ord.User(), req, func(_ comms.Link, msg *msgjson.Message) {
+			m.handlePreimageResp(msg, reqData)
+		}, piTimeout, miss)
```

Comment:
In this case, with a non-nil `err` from `Request`/`RequestWithTimeout`, the `piChan` does not get added to the `preimages` map, so the loop below won't wait on it.  But in general what to do in Request when the user is not found is a good question.  We could also add a call to `expire()` inside `RequestWithTimeout` in case `client == nil`, but sine and error is returned, I think it becomes the callers responsibility to handle whatever is appropriate if the client is not found.